Expand `ErrorHandler` to take more information

jcredding opened the pull request:
This updates the `ErrorHandler` to take more information about the
job that failed. This includes passing the queue that the job
came from, the serialized payload of the job and the handler
class that was being used to run it. The handler class can be `nil`
if the error happened before a handler class was found.

These are being added to expand the functionality of the error
handler and to help with handling errors that happen before a job
is deserialized. Specifically, this is setup for changing how the
hard shutdown feature works. A new version of dat worker pool will
raise an exception on worker threads that are running jobs when it
attempts to force a shutdown. This means an exception can happen
at any point, even before a job has been deserialized. For this
case, we still want to get the serialized payload so we can add
the job back to the queue because it never had a chance to run.
By passing the queue key and serialized payload to the error
handler this will be possible.

This also updates to the latest assert to keep up-to-date and to
use its latest features.

@kellyredding - Ready for review.

kellyredding commented:
@jcredding maybe break up that commit msg into paragraphs for each point of focus or idea.  It kinda runs together as is.

kellyredding commented:
@jcredding "These are being added to expand the functionality..." really starts a new line of thought so it seems logical to para-break there.

kellyredding commented on lib/qs/error_handler.rb:
```diff
@@ -2,12 +2,20 @@ module Qs
 
-    attr_reader :exception, :daemon_data, :job
+    attr_reader :exception, :context
     attr_reader :error_procs
```
@jcredding I like how the error handler API is shaping up: exception, context and error procs - good stuff.  Maybe just use 1 line of readers here now?

kellyredding commented on lib/qs/error_handler.rb:
```diff
@@ -15,11 +23,10 @@ def initialize(exception, daemon_data, job = nil)
       @error_procs.each do |error_proc|
         begin
-          error_proc.call(@exception, @daemon_data, @job)
+          error_proc.call(@exception, @context)
```
@jcredding this is cool - this is all error handling should be about - the exception and any context to it.  good stuff.

kellyredding commented on lib/qs/error_handler.rb:
```diff
@@ -2,12 +2,20 @@ module Qs
+        context[:serialized_payload],
+        context[:job],
+        context[:handler_class]
+      )
```
@jcredding not sure if you want to change this here or not, but it feels like the `ErrorContext` shouldn't be a struct.  It feels like you should just create it with a context hash.  Just a thought since you are manually building it from a hash anyway.

kellyredding commented on lib/qs/error_handler.rb:
```diff
@@ -2,12 +2,20 @@ module Qs
+        context[:job],
+        context[:handler_class]
+      )
+      @error_procs = daemon_data.error_procs.reverse
```
@jcredding you technically don't have to save `daemon_data` off to a local - you could get it from the context here instead.  Just a thought.

kellyredding commented:
@jcredding sorry for the spam just processing the implementation.

One thought is that there is a lot of "ceremony" to getting the context to the error handling procs: the payload handler gathers the context and builds a "processed payload".  Then the on handle error, it deconstructs its processed payload to a hash to create an error handler with just so the error handler can construct its `ErrorContext` from that hash.

You could make the `ErrorContext` a little more formal and have the payload handler build one of those directly and have the `ErrorHandler` just take an error context instance directly?  This would save the hash transition step.

Now, I recognize I don't have the full design/needs of the `ProcessedPayload` and `ErrorContext` in my head and there may be good reasons to keep them abstracted and more dedicated in purpose.  Plus it may introduce too much coupling to take the above approach.  Even if you think it is a good idea, it may not be the right time to switch to it with your other efforts already in the works.  Just wanted to throw it out as a thought though.

kellyredding commented:
@jcredding also, a negative about the approach I mention above is that the hash layer is really flexible and would I think reduce coupling?

kellyredding commented:
@jcredding made a bunch of comments that you may or may not want to address.  Ultimately I'm fine as is on this (barring the commit msg cleanup).  :-1:

jcredding commented:
@kellyredding - I've updated this, fixed the commit message and did some of the things you suggested. 

For now, I'm going to keep the `ErrorContext` internal to keep things decoupled and because I'd rather not add extra logic to the `PayloadHandler` at this point. Basically, I'd rather put the complexity into the error handler instead of the payload handler for now. Also, since the `ErrorContext` is no longer a struct and instead takes a hash, whether I build it internally or externally of the error handler I still have to convert the processed payload to a hash.

kellyredding commented:
@jcredding changes look good - I'm cool with where this is :-1:
